server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -161,6 +167,9 @@ public PluginsService(
 
         this.info = new PluginsAndModules(pluginsList, modulesList);
         this.plugins = Collections.unmodifiableList(pluginsLoaded);
+        List<Plugin> reversePlugins = pluginsLoaded.stream().map(Tuple::v2).collect(Collectors.toCollection(ArrayList::new));
```

Comment:
This completely flips around the hierarchy of extended plugins, and I'm not sure it can work, at least not right now. Let's look at a concrete example of painless spi. Currrently, painless is loaded first, and any plugins implementing painless SPI get loaded after, then a callback is made to tell painless to reload its SPI implementations. In order for the extending plugins to be loaded, painless's SPI jar must be on the classpath. Currently we do this in a lazy way: we actually make painless itself (which includes the spi jar) an "uncle" classloader of each of the extending plugins. If painless spi jar is not already loaded, then the extending plugins could get classnotfoundexception if they reference any of the painless spi classes when loading themselves.  We *could* reverse this order if we separately loaded the spi jars into their own classloaders, and I have long wanted to have them in separate classloaders for other reasons (mainly the possible leakage of eg painless classes/instances into other plugins). But until we do that, I don't think we can even consider reversing the order. Additionally, if we reverse the order, then there is no reason to have the reloadSPI method, as all the SPI implementations would already be available when eg painless is loaded. We would however need to think about how to pass the appropriate classloader instances, although I think this is better done with our own service loader that can be aware of the extended plugin hierarchy.

Reply:
`reversePlugins` is not used for loading (as in new'ing) plugins. They are still loaded in the same order as before this PR and `reloadSPI` is still called in the same order. This is also the order expected by the new `extensionPlugin` method.

However, it is used when calling methods on the plugins. Here we are past all the class loading and it therefore makes most sense to do this in the reverse order. This has to do with how createComponents is used for initializing variables in almost every plugin.

By reversing the order of caling plugin methods, we allow `Autoscaling` to create an object directly that contains all the deciders (extension point implementations) rather than resort to a `Supplier<>` hack.

Reversing the order of loading the plugins might be possible too, but yeah that would require more fundamental changes.

Reply:
> By reversing the order of caling plugin methods, we allow Autoscaling to create an object directly that contains all the deciders (extension point implementations) rather than resort to a Supplier<> hack.

This doesn't make sense to me. reloadSPI is explicitly for this purpose. It gets called as each extended plugin is loaded, and autoscaling is able to check the just loaded plugin for any implementations of its SPI classes, ie deciders.